app/assets/javascripts/listings/components/listing/propertyHero.js.coffee

```diff
@@ -37,5 +37,89 @@ angular.module('dahlia.components')
         index = _.findIndex(listing.reservedDescriptor, ['name', descriptor])
         @reservedUnitIcons[index]
 
+      @_getPercentString = (v) -> "#{v}%"
+
+      @getCurrencyString = (v) -> $filter('currency')(v, '$', 0)
+
+      @getCurrencyRange = (min, max) ->
+        if min? && max?
+          $translate.instant('listings.stats.currency_range', {
+            currencyMinValue: @getCurrencyString(min)
+            currencyMaxValue: @getCurrencyString(max)
+          })
+        else if max?
+          $translate.instant('listings.stats.currency_range', {
+            currencyMinValue: @getCurrencyString(0)
+            currencyMaxValue: @getCurrencyString(max)
+          })
+        else if min?
+          @getCurrencyString(min)
+        else
+          ""
+
+      @getIncomeRangeFromUnitGroup = (unitGroup) -> @getCurrencyRange(unitGroup.minIncome, unitGroup.maxIncome)
+      @getIncomeRangeFromPriceGroup = (priceGroup) -> @getCurrencyRange(priceGroup.minIncome, priceGroup.maxIncome)
+
+      @getHouseholdTextFromOccupancy = (occupancy) ->
+        if occupancy == '1'
+          $translate.instant('listings.stats.num_in_household_singular')
+        else
+          $translate.instant('listings.stats.num_in_household_plural')
+
+      @isWaitlist = (priceGroup) -> priceGroup.Status == "Occupied"
+
+      @numAvailableString = (priceGroup) ->
+        if @isWaitlist(priceGroup)
+          $translate.instant('listings.stats.waitlist')
+        else
+          priceGroup.total + " " + $translate.instant('listings.stats.available')
+
+      occupanciesToIsExpanded = {}
+
+      @isExpanded = (priceGroup) ->
+        hasOnlyOneAccordionGroup = this.parent.listing.groupedUnits.length == 1
+        groupNotToggledYet = occupanciesToIsExpanded[priceGroup.occupancy] == undefined
+
+        # Start with group expanded if there's only one group.
+        if hasOnlyOneAccordionGroup && groupNotToggledYet
+          true
+        else
+          !!occupanciesToIsExpanded[priceGroup.occupancy]
+
+      @toggleExpanded = (priceGroup) ->
+        occupanciesToIsExpanded[priceGroup.occupancy] = !@isExpanded(priceGroup)
+
+      @labelForIncomeLevelTable = (occupancy, incomeLevelString) ->
+        $translate.instant('listings.stats.table_label', {
+          numInHouseholdString: "#{occupancy} #{@getHouseholdTextFromOccupancy(occupancy)}",
+          amiPercentString: incomeLevelString
+        })
+
+      @_rowId = (prefix, unitGroup, incomeLevelIndex, priceGroupIndex) ->
+        "#{prefix}__occupancy_#{unitGroup.occupancy}__income_#{incomeLevelIndex}__price_group_#{priceGroupIndex}"
+
+      @unitTypeId = (unitGroup, incomeLevelIndex, priceGroupIndex) ->
+        @_rowId('unit_type', unitGroup, incomeLevelIndex, priceGroupIndex)
+
+      @incomeRowId = (unitGroup, incomeLevelIndex, priceGroupIndex) ->
+        @_rowId('income', unitGroup, incomeLevelIndex, priceGroupIndex)
```

Comment:
So these actually don't get re-calculated on each render, because we're using the ::$ctrl.incomeRowId() syntax (note the `::`). I think it's okay to leave as-is